### PR TITLE
Add Unit Tests

### PR DESCRIPTION
**Pull request by nilsleh**
This PR aims to add a basic set of unit tests to make the code base more stable.

The overall goal should be to test that:
- the provided configs are enough to instantiate a model wrapper and run a lightning fit call
- automatic download of pretrained paths works
- code modules should be independently usable with or with main.py script

Implementation approach:
- add a dummy datamodule that can generate dummy data for any model
- iterate over `src/configs` to test correctness of configs and also whether modules can be used independently as building blocks

Types of errors encountered so far:
- configs miss attributes that are actually required by the code, this could be handled differently with `dict.get('name', default)` if you want to reduce the number of required params
- import errors because import statements within src do not point to correct places

DO NOT MERGE YET

### CODE REVIEW

**Comment by adamjstewart** [comment]
What's the current coverage percentage?

**Comment by nilsleh** [comment]
Locally, I am currently getting 38% getting for the classification tests I have implemented. Segmentation should add roughly that same number on top. Need to figure out how to get the CLI to run with those installs.

File: src/foundation_models/dofa_wrapper.py [hunk 1]
```diff
@@ -7,16 +7,14 @@
...
-from util.misc import resize
 from .lightning_task import LightningTask
 from timm.models.layers import trunc_normal_
-from util.misc import seg_metric, cls_metric
+from ..util.misc import resize, seg_metric, cls_metric
 from torchvision.datasets.utils import download_url
 from peft import LoraConfig, get_peft_model
 
 
 class DofaClassification(LightningTask):
```
Comment (nilsleh) [inline src/foundation_models/dofa_wrapper.py:19]: I would vote that the DofaModel wrapper functions here and also for all other models, that is just an if/else statement depending on 'classification' or 'segmentation', because via the config files, the respective classes can be targeted directly via the config files.

File: src/foundation_models/senpamae_wrapper.py [hunk 2]
```diff
@@ -15,40 +15,43 @@
 
 
 class SenPaMAEClassification(LightningTask):
-    url = "TBD" #waiting on weights
+    url = "https://drive.google.com/file/d/16IoG47yzdyUnPqUgaV8ofeja5RgQjlAz"
```
Comment (nilsleh) [inline src/foundation_models/senpamae_wrapper.py:19]: Should also be possible to do an automatic download for senpae with torchvision `download_url`

File: src/main.py [hunk 3]
```diff
@@ -29,10 +28,10 @@ def print_trainable_parameters(model):
         f"trainable params: {trainable_params} || all params: {all_param} || trainable%: {100 * trainable_params / all_param:.2f}"
     )
 
+
 @hydra.main(config_path="configs", config_name="config")
 def main(cfg: DictConfig):
-
-    print(f'MODEL_SIZE: {os.getenv("MODEL_SIZE")}')
+    print(f"MODEL_SIZE: {os.getenv('MODEL_SIZE')}")
```
Comment (nilsleh) [inline src/main.py]: I personally don't like the idea of this model size environment variable, because that could just be again configured via the config files? This is also an argument for reproducability. It is more direct, if the model size is not an env variable in the config but concrete.

File: src/foundation_models/softcon_wrapper.py [hunk 4]
```diff
@@ -30,19 +28,22 @@ def __init__(self, args, model_config, data_config):
...
+                    filename=os.path.basename(path),
+                )
+
+            ckpt_vit14 = torch.load(path)
+            self.encoder.load_state_dict(ckpt_vit14)
 
         if model_config.freeze_backbone:
             self.freeze(self.encoder)
 
+        # TODO this embed dim could be pulled from the encoder, to remove the need for the arg
```
Comment (nilsleh) [inline src/foundation_models/softcon_wrapper.py:49]: The embedding dim should be retrieved from the last layer of the encoder, that would remove the need for another config argument.

File: src/foundation_models/CROMA/use_croma.py [hunk 5]
```diff
@@ -20,9 +20,9 @@ def __init__(
         """
         super().__init__()
         # check types
-        assert isinstance(pretrained_path, str), (
-            f"pretrained_path must be a string, not {type(pretrained_path)}"
-        )
+        # assert isinstance(pretrained_path, str), (
```
Comment (nilsleh) [inline src/foundation_models/CROMA/use_croma.py:23]: Could also use random CROMA model and not just pretrained.

**Review by nilsleh — COMMENTED** [review]
@xiong-zhitong @ando-shah @LeWaldm @adamjstewart Here are some comments about things I've found while trying to implement unit tests. Open to discuss any of these points.

Currently test coverage is now 65%, a large missing portion is the ViT-Adapter that is not being tested, because that is also it's own installation issue. Getting the tests to run on CI would be an important next step so we can check the code when different people add parts and we do not break things.

In the long term, we should also remove all the dataset wrappers. The configs should be able to target any datamodule you have locally, or from another library like torchgeo directly. This makes it simpler to use, because there is no need to add a specific dataset wrapper to this repo just to be able to use a dataset of their choice.

File: src/foundation_models/CROMA/use_croma.py (hunk 5 above)
Comment (nilsleh) [reply src/foundation_models/CROMA/use_croma.py:23]: Also probably could remove the entire logic, and just use the torchgeo CROMA model then we only need an import and can remove all this code

File: src/foundation_models/SoftCON/models/dinov2/vision_transformer.py [hunk 6]
```diff
@@ -55,6 +55,7 @@ def forward(self, x):
         return x
 
 
+# TODO is this needed? Or can also be loaded from the HUB?
```
Comment (nilsleh) [inline src/foundation_models/SoftCON/models/dinov2/vision_transformer.py]: @wangyi111 is all of this necessary, and couldn't this just be loaded from torch.hub similar to how the dinov2 wrapper works? Then we could remove this

Comment (wangyi111) [reply src/foundation_models/SoftCON/models/dinov2/vision_transformer.py]: yes it's fine to just load from torch.hub, with something like 
```
model.patch_embed.proj = torch.nn.Conv2d(n_chans, 384, kernel_size=(14, 14), stride=(14, 14)) # n_chans=2 or 13
```
afterwards.

Comment (nilsleh) [reply src/foundation_models/SoftCON/models/dinov2/vision_transformer.py]: okay, removed now in the latest commit, that should make things easier.

File: src/foundation_models/CROMA/use_croma.py (hunk 5 above)
Comment (nilsleh) [reply src/foundation_models/CROMA/use_croma.py:23]: Actually, does not work yet, because in torchgeo the list of output features is not supported, which is needed for segmentation.

File: src/foundation_models/anysat_wrapper.py [hunk 7]
```diff
@@ -4,9 +4,9 @@
...
 from mmseg.models.necks import Feature2Pyramid
 from mmseg.models.decode_heads import UPerHead, FCNHead
-from util.misc import resize
+
 from .lightning_task import LightningTask
-from util.misc import seg_metric, cls_metric
+from ..util.misc import resize, seg_metric, cls_metric
 
 
 class AnySatClassification(LightningTask):
```
Comment (nilsleh) [inline src/foundation_models/anysat_wrapper.py:12]: @xiong-zhitong is Anysat correctly configured for segmentation? based on the notebook: https://github.com/gastruc/AnySat/blob/main/demo.ipynb it seems the model does not produce a list of features only a fixed feature output, so not sure how that works with the neck rescales

File: src/configs/config.yaml [hunk 8]
```diff
@@ -9,13 +9,14 @@ lr: 1e-3
 batch_size: 32
 weight_decay: 0.0001
 num_workers: 8
-pin_mem: true
-num_gpus: 2
-strategy: ddp
+# pin_mem: true
```
Comment (nilsleh) [inline src/configs/config.yaml]: Also I don't think any of these should be specified in the default global dict? or at least should be set to default null? I don't think we should assume the default strategy to be ddp

Comment (nilsleh) [reply src/configs/config.yaml]: I think we should probably make a clear outline of what is a model_config, data_config and additional parameters

**Comment by nilsleh** [comment]
The majority of datasets are missing wavelenght parameters, which are necessary to run DOFA

File: src/main.py (hunk 3 above)
Comment (LeWaldm) [reply src/main.py]: I agree, I found the env variable counter intuitive. Having everything in the config files sounds more reproducible to me.

**Comment by adamjstewart** [comment]
For now we can add wavelength info via our wrapper datasets, but @wangyi111 and I are slowly adding this info to the underlying TorchGeo datasets.

File: src/configs/config.yaml (hunk 8 above)
Comment (ando-shah) [reply src/configs/config.yaml]: I agree. num_gpus is currently being set by the shell script calling main

Comment (adamjstewart) [reply src/configs/config.yaml]: Can't we just use Lightning instead?

Comment (nilsleh) [reply src/configs/config.yaml]: It is used in lightning: https://github.com/xiong-zhitong/DOFA-pytorch/blob/b60fb43f35a1283acdf262d75b15da7b44e450eb/src/main.py#L72

Comment (adamjstewart) [reply src/configs/config.yaml]: *Can't we just use LightningCLI instead?

Comment (nilsleh) [reply src/configs/config.yaml]: You could, but I don't think we have time to configure that properly with the current setup of the code base.
